// @turf/turf/dist/js/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var isolines = require('@turf/isolines');
var convex = require('@turf/convex');
var pointsWithinPolygon = require('@turf/points-within-polygon');
var concave = require('@turf/concave');
var collect = require('@turf/collect');
var flip = require('@turf/flip');
var simplify = require('@turf/simplify');
var bezierSpline = require('@turf/bezier-spline');
var tag = require('@turf/tag');
var sample = require('@turf/sample');
var envelope = require('@turf/envelope');
var square = require('@turf/square');
var circle = require('@turf/circle');
var midpoint = require('@turf/midpoint');
var center = require('@turf/center');
var centerOfMass = require('@turf/center-of-mass');
var centroid = require('@turf/centroid');
var combine = require('@turf/combine');
var distance = require('@turf/distance');
var explode = require('@turf/explode');
var bbox = require('@turf/bbox');
var tesselate = require('@turf/tesselate');
var bboxPolygon = require('@turf/bbox-polygon');
var booleanPointInPolygon = require('@turf/boolean-point-in-polygon');
var nearestPoint = require('@turf/nearest-point');
var nearestPointOnLine = require('@turf/nearest-point-on-line');
var nearestPointToLine = require('@turf/nearest-point-to-line');
var planepoint = require('@turf/planepoint');
var tin = require('@turf/tin');
var bearing = require('@turf/bearing');
var destination = require('@turf/destination');
var kinks = require('@turf/kinks');
var pointOnFeature = require('@turf/point-on-feature');
var area = require('@turf/area');
var along = require('@turf/along');
var length = require('@turf/length');
var lineSlice = require('@turf/line-slice');
var lineSliceAlong = require('@turf/line-slice-along');
var pointGrid = require('@turf/point-grid');
var truncate = require('@turf/truncate');
var flatten = require('@turf/flatten');
var lineIntersect = require('@turf/line-intersect');
var lineChunk = require('@turf/line-chunk');
var unkinkPolygon = require('@turf/unkink-polygon');
var greatCircle = require('@turf/great-circle');
var lineSegment = require('@turf/line-segment');
var lineSplit = require('@turf/line-split');
var lineArc = require('@turf/line-arc');
var polygonToLine = require('@turf/polygon-to-line');
var lineToPolygon = require('@turf/line-to-polygon');
var bboxClip = require('@turf/bbox-clip');
var lineOverlap = require('@turf/line-overlap');
var sector = require('@turf/sector');
var rhumbBearing = require('@turf/rhumb-bearing');
var rhumbDistance = require('@turf/rhumb-distance');
var rhumbDestination = require('@turf/rhumb-destination');
var polygonTangents = require('@turf/polygon-tangents');
var rewind = require('@turf/rewind');
var isobands = require('@turf/isobands');
var transformRotate = require('@turf/transform-rotate');
var transformScale = require('@turf/transform-scale');
var transformTranslate = require('@turf/transform-translate');
var lineOffset = require('@turf/line-offset');
var polygonize = require('@turf/polygonize');
var booleanDisjoint = require('@turf/boolean-disjoint');
var booleanContains = require('@turf/boolean-contains');
var booleanCrosses = require('@turf/boolean-crosses');
var booleanClockwise = require('@turf/boolean-clockwise');
var booleanOverlap = require('@turf/boolean-overlap');
var booleanPointOnLine = require('@turf/boolean-point-on-line');
var booleanEqual = require('@turf/boolean-equal');
var booleanWithin = require('@turf/boolean-within');
var booleanIntersects = require('@turf/boolean-intersects');
var clone = require('@turf/clone');
var cleanCoords = require('@turf/clean-coords');
var clustersDbscan = require('@turf/clusters-dbscan');
var clustersKmeans = require('@turf/clusters-kmeans');
var pointToLineDistance = require('@turf/point-to-line-distance');
var booleanParallel = require('@turf/boolean-parallel');
var shortestPath = require('@turf/shortest-path');
var voronoi = require('@turf/voronoi');
var ellipse = require('@turf/ellipse');
var centerMean = require('@turf/center-mean');
var centerMedian = require('@turf/center-median');
var standardDeviationalEllipse = require('@turf/standard-deviational-ellipse');
var angle = require('@turf/angle');
var polygonSmooth = require('@turf/polygon-smooth');
var moranIndex = require('@turf/moran-index');
var distanceWeight = require('@turf/distance-weight');
var projection = require('@turf/projection');
var random = require('@turf/random');
var clusters = require('@turf/clusters');
var helpers = require('@turf/helpers');
var invariant = require('@turf/invariant');
var meta = require('@turf/meta');
var difference = require('@turf/difference');
var buffer = require('@turf/buffer');
var union = require('@turf/union');
var intersect = require('@turf/intersect');
var dissolve = require('@turf/dissolve');
var hexGrid = require('@turf/hex-grid');
var mask = require('@turf/mask');
var squareGrid = require('@turf/square-grid');
var triangleGrid = require('@turf/triangle-grid');
var interpolate = require('@turf/interpolate');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () {
            return e[k];
          }
        });
      }
    });
  }
  n['default'] = e;
  return Object.freeze(n);
}

var isolines__default = /*#__PURE__*/_interopDefaultLegacy(isolines);
var convex__default = /*#__PURE__*/_interopDefaultLegacy(convex);
var pointsWithinPolygon__default = /*#__PURE__*/_interopDefaultLegacy(pointsWithinPolygon);
var concave__default = /*#__PURE__*/_interopDefaultLegacy(concave);
var collect__default = /*#__PURE__*/_interopDefaultLegacy(collect);
var flip__default = /*#__PURE__*/_interopDefaultLegacy(flip);
var simplify__default = /*#__PURE__*/_interopDefaultLegacy(simplify);
var bezierSpline__default = /*#__PURE__*/_interopDefaultLegacy(bezierSpline);
var tag__default = /*#__PURE__*/_interopDefaultLegacy(tag);
var sample__default = /*#__PURE__*/_interopDefaultLegacy(sample);
var envelope__default = /*#__PURE__*/_interopDefaultLegacy(envelope);
var square__default = /*#__PURE__*/_interopDefaultLegacy(square);
var circle__default = /*#__PURE__*/_interopDefaultLegacy(circle);
var midpoint__default = /*#__PURE__*/_interopDefaultLegacy(midpoint);
var center__default = /*#__PURE__*/_interopDefaultLegacy(center);
var centerOfMass__default = /*#__PURE__*/_interopDefaultLegacy(centerOfMass);
var centroid__default = /*#__PURE__*/_interopDefaultLegacy(centroid);
var combine__default = /*#__PURE__*/_interopDefaultLegacy(combine);
var distance__default = /*#__PURE__*/_interopDefaultLegacy(distance);
var explode__default = /*#__PURE__*/_interopDefaultLegacy(explode);
var bbox__default = /*#__PURE__*/_interopDefaultLegacy(bbox);
var tesselate__default = /*#__PURE__*/_interopDefaultLegacy(tesselate);
var bboxPolygon__default = /*#__PURE__*/_interopDefaultLegacy(bboxPolygon);
var booleanPointInPolygon__default = /*#__PURE__*/_interopDefaultLegacy(booleanPointInPolygon);
var nearestPoint__default = /*#__PURE__*/_interopDefaultLegacy(nearestPoint);
var nearestPointOnLine__default = /*#__PURE__*/_interopDefaultLegacy(nearestPointOnLine);
var nearestPointToLine__default = /*#__PURE__*/_interopDefaultLegacy(nearestPointToLine);
var planepoint__default = /*#__PURE__*/_interopDefaultLegacy(planepoint);
var tin__default = /*#__PURE__*/_interopDefaultLegacy(tin);
var bearing__default = /*#__PURE__*/_interopDefaultLegacy(bearing);
var destination__default = /*#__PURE__*/_interopDefaultLegacy(destination);
var kinks__default = /*#__PURE__*/_interopDefaultLegacy(kinks);
var pointOnFeature__default = /*#__PURE__*/_interopDefaultLegacy(pointOnFeature);
var area__default = /*#__PURE__*/_interopDefaultLegacy(area);
var along__default = /*#__PURE__*/_interopDefaultLegacy(along);
var length__default = /*#__PURE__*/_interopDefaultLegacy(length);
var lineSlice__default = /*#__PURE__*/_interopDefaultLegacy(lineSlice);
var lineSliceAlong__default = /*#__PURE__*/_interopDefaultLegacy(lineSliceAlong);
var pointGrid__default = /*#__PURE__*/_interopDefaultLegacy(pointGrid);
var truncate__default = /*#__PURE__*/_interopDefaultLegacy(truncate);
var flatten__default = /*#__PURE__*/_interopDefaultLegacy(flatten);
var lineIntersect__default = /*#__PURE__*/_interopDefaultLegacy(lineIntersect);
var lineChunk__default = /*#__PURE__*/_interopDefaultLegacy(lineChunk);
var unkinkPolygon__default = /*#__PURE__*/_interopDefaultLegacy(unkinkPolygon);
var greatCircle__default = /*#__PURE__*/_interopDefaultLegacy(greatCircle);
var lineSegment__default = /*#__PURE__*/_interopDefaultLegacy(lineSegment);
var lineSplit__default = /*#__PURE__*/_interopDefaultLegacy(lineSplit);
var lineArc__default = /*#__PURE__*/_interopDefaultLegacy(lineArc);
var polygonToLine__default = /*#__PURE__*/_interopDefaultLegacy(polygonToLine);
var lineToPolygon__default = /*#__PURE__*/_interopDefaultLegacy(lineToPolygon);
var bboxClip__default = /*#__PURE__*/_interopDefaultLegacy(bboxClip);
var lineOverlap__default = /*#__PURE__*/_interopDefaultLegacy(lineOverlap);
var sector__default = /*#__PURE__*/_interopDefaultLegacy(sector);
var rhumbBearing__default = /*#__PURE__*/_interopDefaultLegacy(rhumbBearing);
var rhumbDistance__default = /*#__PURE__*/_interopDefaultLegacy(rhumbDistance);
var rhumbDestination__default = /*#__PURE__*/_interopDefaultLegacy(rhumbDestination);
var polygonTangents__default = /*#__PURE__*/_interopDefaultLegacy(polygonTangents);
var rewind__default = /*#__PURE__*/_interopDefaultLegacy(rewind);
var isobands__default = /*#__PURE__*/_interopDefaultLegacy(isobands);
var transformRotate__default = /*#__PURE__*/_interopDefaultLegacy(transformRotate);
var transformScale__default = /*#__PURE__*/_interopDefaultLegacy(transformScale);
var transformTranslate__default = /*#__PURE__*/_interopDefaultLegacy(transformTranslate);
var lineOffset__default = /*#__PURE__*/_interopDefaultLegacy(lineOffset);
var polygonize__default = /*#__PURE__*/_interopDefaultLegacy(polygonize);
var booleanDisjoint__default = /*#__PURE__*/_interopDefaultLegacy(booleanDisjoint);
var booleanContains__default = /*#__PURE__*/_interopDefaultLegacy(booleanContains);
var booleanCrosses__default = /*#__PURE__*/_interopDefaultLegacy(booleanCrosses);
var booleanClockwise__default = /*#__PURE__*/_interopDefaultLegacy(booleanClockwise);
var booleanOverlap__default = /*#__PURE__*/_interopDefaultLegacy(booleanOverlap);
var booleanPointOnLine__default = /*#__PURE__*/_interopDefaultLegacy(booleanPointOnLine);
var booleanEqual__default = /*#__PURE__*/_interopDefaultLegacy(booleanEqual);
var booleanWithin__default = /*#__PURE__*/_interopDefaultLegacy(booleanWithin);
var booleanIntersects__default = /*#__PURE__*/_interopDefaultLegacy(booleanIntersects);
var clone__default = /*#__PURE__*/_interopDefaultLegacy(clone);
var cleanCoords__default = /*#__PURE__*/_interopDefaultLegacy(cleanCoords);
var clustersDbscan__default = /*#__PURE__*/_interopDefaultLegacy(clustersDbscan);
var clustersKmeans__default = /*#__PURE__*/_interopDefaultLegacy(clustersKmeans);
var pointToLineDistance__default = /*#__PURE__*/_interopDefaultLegacy(pointToLineDistance);
var booleanParallel__default = /*#__PURE__*/_interopDefaultLegacy(booleanParallel);
var shortestPath__default = /*#__PURE__*/_interopDefaultLegacy(shortestPath);
var voronoi__default = /*#__PURE__*/_interopDefaultLegacy(voronoi);
var ellipse__default = /*#__PURE__*/_interopDefaultLegacy(ellipse);
var centerMean__default = /*#__PURE__*/_interopDefaultLegacy(centerMean);
var centerMedian__default = /*#__PURE__*/_interopDefaultLegacy(centerMedian);
var standardDeviationalEllipse__default = /*#__PURE__*/_interopDefaultLegacy(standardDeviationalEllipse);
var angle__default = /*#__PURE__*/_interopDefaultLegacy(angle);
var polygonSmooth__default = /*#__PURE__*/_interopDefaultLegacy(polygonSmooth);
var moranIndex__default = /*#__PURE__*/_interopDefaultLegacy(moranIndex);
var distanceWeight__default = /*#__PURE__*/_interopDefaultLegacy(distanceWeight);
var projection__namespace = /*#__PURE__*/_interopNamespace(projection);
var random__namespace = /*#__PURE__*/_interopNamespace(random);
var clusters__namespace = /*#__PURE__*/_interopNamespace(clusters);
var helpers__namespace = /*#__PURE__*/_interopNamespace(helpers);
var invariant__namespace = /*#__PURE__*/_interopNamespace(invariant);
var meta__namespace = /*#__PURE__*/_interopNamespace(meta);
var difference__default = /*#__PURE__*/_interopDefaultLegacy(difference);
var buffer__default = /*#__PURE__*/_interopDefaultLegacy(buffer);
var union__default = /*#__PURE__*/_interopDefaultLegacy(union);
var intersect__default = /*#__PURE__*/_interopDefaultLegacy(intersect);
var dissolve__default = /*#__PURE__*/_interopDefaultLegacy(dissolve);
var hexGrid__default = /*#__PURE__*/_interopDefaultLegacy(hexGrid);
var mask__default = /*#__PURE__*/_interopDefaultLegacy(mask);
var squareGrid__default = /*#__PURE__*/_interopDefaultLegacy(squareGrid);
var triangleGrid__default = /*#__PURE__*/_interopDefaultLegacy(triangleGrid);
var interpolate__default = /*#__PURE__*/_interopDefaultLegacy(interpolate);



Object.keys(projection).forEach(function (k) {
  if (k !== 'default') Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return projection[k];
    }
  });
});
Object.keys(random).forEach(function (k) {
  if (k !== 'default') Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return random[k];
    }
  });
});
Object.keys(clusters).forEach(function (k) {
  if (k !== 'default') Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return clusters[k];
    }
  });
});
Object.keys(helpers).forEach(function (k) {
  if (k !== 'default') Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return helpers[k];
    }
  });
});
Object.keys(invariant).forEach(function (k) {
  if (k !== 'default') Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return invariant[k];
    }
  });
});
Object.keys(meta).forEach(function (k) {
  if (k !== 'default') Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return meta[k];
    }
  });
});
Object.defineProperty(exports, 'isolines', {
  enumerable: true,
  get: function () {
    return isolines__default['default'];
  }
});
Object.defineProperty(exports, 'convex', {
  enumerable: true,
  get: function () {
    return convex__default['default'];
  }
});
Object.defineProperty(exports, 'pointsWithinPolygon', {
  enumerable: true,
  get: function () {
    return pointsWithinPolygon__default['default'];
  }
});
Object.defineProperty(exports, 'within', {
  enumerable: true,
  get: function () {
    return pointsWithinPolygon__default['default'];
  }
});
Object.defineProperty(exports, 'concave', {
  enumerable: true,
  get: function () {
    return concave__default['default'];
  }
});
Object.defineProperty(exports, 'collect', {
  enumerable: true,
  get: function () {
    return collect__default['default'];
  }
});
Object.defineProperty(exports, 'flip', {
  enumerable: true,
  get: function () {
    return flip__default['default'];
  }
});
Object.defineProperty(exports, 'simplify', {
  enumerable: true,
  get: function () {
    return simplify__default['default'];
  }
});
Object.defineProperty(exports, 'bezier', {
  enumerable: true,
  get: function () {
    return bezierSpline__default['default'];
  }
});
Object.defineProperty(exports, 'bezierSpline', {
  enumerable: true,
  get: function () {
    return bezierSpline__default['default'];
  }
});
Object.defineProperty(exports, 'tag', {
  enumerable: true,
  get: function () {
    return tag__default['default'];
  }
});
Object.defineProperty(exports, 'sample', {
  enumerable: true,
  get: function () {
    return sample__default['default'];
  }
});
Object.defineProperty(exports, 'envelope', {
  enumerable: true,
  get: function () {
    return envelope__default['default'];
  }
});
Object.defineProperty(exports, 'square', {
  enumerable: true,
  get: function () {
    return square__default['default'];
  }
});
Object.defineProperty(exports, 'circle', {
  enumerable: true,
  get: function () {
    return circle__default['default'];
  }
});
Object.defineProperty(exports, 'midpoint', {
  enumerable: true,
  get: function () {
    return midpoint__default['default'];
  }
});
Object.defineProperty(exports, 'center', {
  enumerable: true,
  get: function () {
    return center__default['default'];
  }
});
Object.defineProperty(exports, 'centerOfMass', {
  enumerable: true,
  get: function () {
    return centerOfMass__default['default'];
  }
});
Object.defineProperty(exports, 'centroid', {
  enumerable: true,
  get: function () {
    return centroid__default['default'];
  }
});
Object.defineProperty(exports, 'combine', {
  enumerable: true,
  get: function () {
    return combine__default['default'];
  }
});
Object.defineProperty(exports, 'distance', {
  enumerable: true,
  get: function () {
    return distance__default['default'];
  }
});
Object.defineProperty(exports, 'explode', {
  enumerable: true,
  get: function () {
    return explode__default['default'];
  }
});
Object.defineProperty(exports, 'bbox', {
  enumerable: true,
  get: function () {
    return bbox__default['default'];
  }
});
Object.defineProperty(exports, 'tesselate', {
  enumerable: true,
  get: function () {
    return tesselate__default['default'];
  }
});
Object.defineProperty(exports, 'bboxPolygon', {
  enumerable: true,
  get: function () {
    return bboxPolygon__default['default'];
  }
});
Object.defineProperty(exports, 'booleanPointInPolygon', {
  enumerable: true,
  get: function () {
    return booleanPointInPolygon__default['default'];
  }
});
Object.defineProperty(exports, 'inside', {
  enumerable: true,
  get: function () {
    return booleanPointInPolygon__default['default'];
  }
});
Object.defineProperty(exports, 'nearest', {
  enumerable: true,
  get: function () {
    return nearestPoint__default['default'];
  }
});
Object.defineProperty(exports, 'nearestPoint', {
  enumerable: true,
  get: function () {
    return nearestPoint__default['default'];
  }
});
Object.defineProperty(exports, 'nearestPointOnLine', {
  enumerable: true,
  get: function () {
    return nearestPointOnLine__default['default'];
  }
});
Object.defineProperty(exports, 'pointOnLine', {
  enumerable: true,
  get: function () {
    return nearestPointOnLine__default['default'];
  }
});
Object.defineProperty(exports, 'nearestPointToLine', {
  enumerable: true,
  get: function () {
    return nearestPointToLine__default['default'];
  }
});
Object.defineProperty(exports, 'planepoint', {
  enumerable: true,
  get: function () {
    return planepoint__default['default'];
  }
});
Object.defineProperty(exports, 'tin', {
  enumerable: true,
  get: function () {
    return tin__default['default'];
  }
});
Object.defineProperty(exports, 'bearing', {
  enumerable: true,
  get: function () {
    return bearing__default['default'];
  }
});
Object.defineProperty(exports, 'destination', {
  enumerable: true,
  get: function () {
    return destination__default['default'];
  }
});
Object.defineProperty(exports, 'kinks', {
  enumerable: true,
  get: function () {
    return kinks__default['default'];
  }
});
Object.defineProperty(exports, 'pointOnFeature', {
  enumerable: true,
  get: function () {
    return pointOnFeature__default['default'];
  }
});
Object.defineProperty(exports, 'pointOnSurface', {
  enumerable: true,
  get: function () {
    return pointOnFeature__default['default'];
  }
});
Object.defineProperty(exports, 'area', {
  enumerable: true,
  get: function () {
    return area__default['default'];
  }
});
Object.defineProperty(exports, 'along', {
  enumerable: true,
  get: function () {
    return along__default['default'];
  }
});
Object.defineProperty(exports, 'length', {
  enumerable: true,
  get: function () {
    return length__default['default'];
  }
});
Object.defineProperty(exports, 'lineDistance', {
  enumerable: true,
  get: function () {
    return length__default['default'];
  }
});
Object.defineProperty(exports, 'lineSlice', {
  enumerable: true,
  get: function () {
    return lineSlice__default['default'];
  }
});
Object.defineProperty(exports, 'lineSliceAlong', {
  enumerable: true,
  get: function () {
    return lineSliceAlong__default['default'];
  }
});
Object.defineProperty(exports, 'pointGrid', {
  enumerable: true,
  get: function () {
    return pointGrid__default['default'];
  }
});
Object.defineProperty(exports, 'truncate', {
  enumerable: true,
  get: function () {
    return truncate__default['default'];
  }
});
Object.defineProperty(exports, 'flatten', {
  enumerable: true,
  get: function () {
    return flatten__default['default'];
  }
});
Object.defineProperty(exports, 'lineIntersect', {
  enumerable: true,
  get: function () {
    return lineIntersect__default['default'];
  }
});
Object.defineProperty(exports, 'lineChunk', {
  enumerable: true,
  get: function () {
    return lineChunk__default['default'];
  }
});
Object.defineProperty(exports, 'unkinkPolygon', {
  enumerable: true,
  get: function () {
    return unkinkPolygon__default['default'];
  }
});
Object.defineProperty(exports, 'greatCircle', {
  enumerable: true,
  get: function () {
    return greatCircle__default['default'];
  }
});
Object.defineProperty(exports, 'lineSegment', {
  enumerable: true,
  get: function () {
    return lineSegment__default['default'];
  }
});
Object.defineProperty(exports, 'lineSplit', {
  enumerable: true,
  get: function () {
    return lineSplit__default['default'];
  }
});
Object.defineProperty(exports, 'lineArc', {
  enumerable: true,
  get: function () {
    return lineArc__default['default'];
  }
});
Object.defineProperty(exports, 'polygonToLine', {
  enumerable: true,
  get: function () {
    return polygonToLine__default['default'];
  }
});
Object.defineProperty(exports, 'polygonToLineString', {
  enumerable: true,
  get: function () {
    return polygonToLine__default['default'];
  }
});
Object.defineProperty(exports, 'lineStringToPolygon', {
  enumerable: true,
  get: function () {
    return lineToPolygon__default['default'];
  }
});
Object.defineProperty(exports, 'lineToPolygon', {
  enumerable: true,
  get: function () {
    return lineToPolygon__default['default'];
  }
});
Object.defineProperty(exports, 'bboxClip', {
  enumerable: true,
  get: function () {
    return bboxClip__default['default'];
  }
});
Object.defineProperty(exports, 'lineOverlap', {
  enumerable: true,
  get: function () {
    return lineOverlap__default['default'];
  }
});
Object.defineProperty(exports, 'sector', {
  enumerable: true,
  get: function () {
    return sector__default['default'];
  }
});
Object.defineProperty(exports, 'rhumbBearing', {
  enumerable: true,
  get: function () {
    return rhumbBearing__default['default'];
  }
});
Object.defineProperty(exports, 'rhumbDistance', {
  enumerable: true,
  get: function () {
    return rhumbDistance__default['default'];
  }
});
Object.defineProperty(exports, 'rhumbDestination', {
  enumerable: true,
  get: function () {
    return rhumbDestination__default['default'];
  }
});
Object.defineProperty(exports, 'polygonTangents', {
  enumerable: true,
  get: function () {
    return polygonTangents__default['default'];
  }
});
Object.defineProperty(exports, 'rewind', {
  enumerable: true,
  get: function () {
    return rewind__default['default'];
  }
});
Object.defineProperty(exports, 'isobands', {
  enumerable: true,
  get: function () {
    return isobands__default['default'];
  }
});
Object.defineProperty(exports, 'transformRotate', {
  enumerable: true,
  get: function () {
    return transformRotate__default['default'];
  }
});
Object.defineProperty(exports, 'transformScale', {
  enumerable: true,
  get: function () {
    return transformScale__default['default'];
  }
});
Object.defineProperty(exports, 'transformTranslate', {
  enumerable: true,
  get: function () {
    return transformTranslate__default['default'];
  }
});
Object.defineProperty(exports, 'lineOffset', {
  enumerable: true,
  get: function () {
    return lineOffset__default['default'];
  }
});
Object.defineProperty(exports, 'polygonize', {
  enumerable: true,
  get: function () {
    return polygonize__default['default'];
  }
});
Object.defineProperty(exports, 'booleanDisjoint', {
  enumerable: true,
  get: function () {
    return booleanDisjoint__default['default'];
  }
});
Object.defineProperty(exports, 'booleanContains', {
  enumerable: true,
  get: function () {
    return booleanContains__default['default'];
  }
});
Object.defineProperty(exports, 'booleanCrosses', {
  enumerable: true,
  get: function () {
    return booleanCrosses__default['default'];
  }
});
Object.defineProperty(exports, 'booleanClockwise', {
  enumerable: true,
  get: function () {
    return booleanClockwise__default['default'];
  }
});
Object.defineProperty(exports, 'booleanOverlap', {
  enumerable: true,
  get: function () {
    return booleanOverlap__default['default'];
  }
});
Object.defineProperty(exports, 'booleanPointOnLine', {
  enumerable: true,
  get: function () {
    return booleanPointOnLine__default['default'];
  }
});
Object.defineProperty(exports, 'booleanEqual', {
  enumerable: true,
  get: function () {
    return booleanEqual__default['default'];
  }
});
Object.defineProperty(exports, 'booleanWithin', {
  enumerable: true,
  get: function () {
    return booleanWithin__default['default'];
  }
});
Object.defineProperty(exports, 'booleanIntersects', {
  enumerable: true,
  get: function () {
    return booleanIntersects__default['default'];
  }
});
Object.defineProperty(exports, 'clone', {
  enumerable: true,
  get: function () {
    return clone__default['default'];
  }
});
Object.defineProperty(exports, 'cleanCoords', {
  enumerable: true,
  get: function () {
    return cleanCoords__default['default'];
  }
});
Object.defineProperty(exports, 'clustersDbscan', {
  enumerable: true,
  get: function () {
    return clustersDbscan__default['default'];
  }
});
Object.defineProperty(exports, 'clustersKmeans', {
  enumerable: true,
  get: function () {
    return clustersKmeans__default['default'];
  }
});
Object.defineProperty(exports, 'pointToLineDistance', {
  enumerable: true,
  get: function () {
    return pointToLineDistance__default['default'];
  }
});
Object.defineProperty(exports, 'booleanParallel', {
  enumerable: true,
  get: function () {
    return booleanParallel__default['default'];
  }
});
Object.defineProperty(exports, 'shortestPath', {
  enumerable: true,
  get: function () {
    return shortestPath__default['default'];
  }
});
Object.defineProperty(exports, 'voronoi', {
  enumerable: true,
  get: function () {
    return voronoi__default['default'];
  }
});
Object.defineProperty(exports, 'ellipse', {
  enumerable: true,
  get: function () {
    return ellipse__default['default'];
  }
});
Object.defineProperty(exports, 'centerMean', {
  enumerable: true,
  get: function () {
    return centerMean__default['default'];
  }
});
Object.defineProperty(exports, 'centerMedian', {
  enumerable: true,
  get: function () {
    return centerMedian__default['default'];
  }
});
Object.defineProperty(exports, 'standardDeviationalEllipse', {
  enumerable: true,
  get: function () {
    return standardDeviationalEllipse__default['default'];
  }
});
Object.defineProperty(exports, 'angle', {
  enumerable: true,
  get: function () {
    return angle__default['default'];
  }
});
Object.defineProperty(exports, 'polygonSmooth', {
  enumerable: true,
  get: function () {
    return polygonSmooth__default['default'];
  }
});
Object.defineProperty(exports, 'moranIndex', {
  enumerable: true,
  get: function () {
    return moranIndex__default['default'];
  }
});
Object.defineProperty(exports, 'distanceWeight', {
  enumerable: true,
  get: function () {
    return distanceWeight__default['default'];
  }
});
exports.projection = projection__namespace;
exports.random = random__namespace;
exports.clusters = clusters__namespace;
Object.defineProperty(exports, 'bearingToAngle', {
  enumerable: true,
  get: function () {
    return helpers.bearingToAzimuth;
  }
});
Object.defineProperty(exports, 'convertDistance', {
  enumerable: true,
  get: function () {
    return helpers.convertLength;
  }
});
Object.defineProperty(exports, 'degrees2radians', {
  enumerable: true,
  get: function () {
    return helpers.degreesToRadians;
  }
});
Object.defineProperty(exports, 'distanceToDegrees', {
  enumerable: true,
  get: function () {
    return helpers.lengthToDegrees;
  }
});
Object.defineProperty(exports, 'distanceToRadians', {
  enumerable: true,
  get: function () {
    return helpers.lengthToRadians;
  }
});
exports.helpers = helpers__namespace;
Object.defineProperty(exports, 'radians2degrees', {
  enumerable: true,
  get: function () {
    return helpers.radiansToDegrees;
  }
});
Object.defineProperty(exports, 'radiansToDistance', {
  enumerable: true,
  get: function () {
    return helpers.radiansToLength;
  }
});
exports.invariant = invariant__namespace;
exports.meta = meta__namespace;
Object.defineProperty(exports, 'difference', {
  enumerable: true,
  get: function () {
    return difference__default['default'];
  }
});
Object.defineProperty(exports, 'buffer', {
  enumerable: true,
  get: function () {
    return buffer__default['default'];
  }
});
Object.defineProperty(exports, 'union', {
  enumerable: true,
  get: function () {
    return union__default['default'];
  }
});
Object.defineProperty(exports, 'intersect', {
  enumerable: true,
  get: function () {
    return intersect__default['default'];
  }
});
Object.defineProperty(exports, 'dissolve', {
  enumerable: true,
  get: function () {
    return dissolve__default['default'];
  }
});
Object.defineProperty(exports, 'hexGrid', {
  enumerable: true,
  get: function () {
    return hexGrid__default['default'];
  }
});
Object.defineProperty(exports, 'mask', {
  enumerable: true,
  get: function () {
    return mask__default['default'];
  }
});
Object.defineProperty(exports, 'squareGrid', {
  enumerable: true,
  get: function () {
    return squareGrid__default['default'];
  }
});
Object.defineProperty(exports, 'triangleGrid', {
  enumerable: true,
  get: function () {
    return triangleGrid__default['default'];
  }
});
Object.defineProperty(exports, 'interpolate', {
  enumerable: true,
  get: function () {
    return interpolate__default['default'];
  }
});
